bundles/org.smarthomej.binding.androiddebugbridge/src/main/java/org/smarthomej/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java

```diff
@@ -273,25 +299,32 @@ private String runAdbShell(String... args)
         if (adb == null) {
             throw new AndroidDebugBridgeDeviceException("Device not connected");
         }
-        var commandFuture = scheduler.submit(() -> {
-            var byteArrayOutputStream = new ByteArrayOutputStream();
-            String cmd = String.join(" ", args);
-            logger.debug("{} - shell:{}", ip, cmd);
-            try {
-                AdbStream stream = adb.open("shell:" + cmd);
-                do {
-                    byteArrayOutputStream.writeBytes(stream.read());
-                } while (!stream.isClosed());
-            } catch (IOException e) {
-                String message = e.getMessage();
-                if (message != null && !message.equals("Stream closed")) {
-                    throw e;
+        lock.lock();
+        try {
+            commandFuture = scheduler.submit(() -> {
```

Comment:
Is this still needed as a field in the class if there is a lock?

Reply:
yes because the disconnect method needs to cancel it as soon at it is called. if the thread timeout for example in 30 seconds and the disconnect method is called in the 7th second it has to clean up the thread now. else the thread will be alive for 23 seconds more and this could maybe produce a schedule overflow.

Reply:
Woudl it make sense to check if a `commandFuture` is already set (check for null) and set it to `null` in the `finally` clause?

Reply:
Additional to the disconnect method or as a replacement for it? I think after a cancel a future should always set to null but in the aec-binding we also only set it on the logout method to null, so i did it the same way in the adb-binding. I still don't know if there is any difference but i hope the garbage collector will collect it in any way (is null or thread finished/canceled) I only can say that if an timeout exception appears in the thread because the get method timed out, the thread is still alive and is in a strange condition. If this happens multiple times it seems the max threads of scheduler is reached and no command/thread is executed anymore. This is why we have to cancel it in the finally block, prophylactically. This will stop the scheduler-overflow (same i recognized on the aec-binding before you changed it like it is today). And in the disconnect method like i explained before.

Reply:
The garbage collector will collect it if there is no reference to the object, so it'll be collected. But If the timer was not finished before it is replaced, the timer will execute at the time set before. So generally before assigning a new value it is a good idea to check if there is already a scheduled job.  I would recommend a method like this:

```
private void stopCommandFuture() {
    Future<?> f = commandFuture;
    if (f != null) {
        f.cancel(true);
        commandFuture = null;
    }
 }
```

and call this in all places where you want to end the current job (e.g. in the `finally` clause instead of l. 322-324 and in `disconnect` instead of l. 356-360) and also in all places where you assign a new instance to the `commandFuture` field (before l. 304) . This ensures that at no time there is more than one command future in place.

Regarding the amazonechocontrol: There is no need to set it to `null`, because (e.g. the annoucnement-timer) removed from the `timers`-map here: 

https://github.com/smarthomej/addons/blob/28d6b1f6fff9acfd3749e3634fe369e345cc061c/bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/Connection.java#L1365 

That is essentially the same as setting it to null, because further `get` calls to the map will result in `null`.  New values are only assigned if there is no job currently scheduled:

https://github.com/smarthomej/addons/blob/28d6b1f6fff9acfd3749e3634fe369e345cc061c/bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/Connection.java#L1326-L1327